core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+    private final Map<PlanFragmentId, StageExecution> stagesByFragmentId;
+    private final Set<StageExecution> activeSources = new HashSet<>();
+
+    @GuardedBy("this")
+    private SettableFuture<Void> rescheduleFuture = SettableFuture.create();
+
+    public static SequentialExecutionSchedule forStages(Collection<PipelinedStageExecution> stages)
+    {
+        return new SequentialExecutionSchedule(stages.stream()
+                .map(SequentialExecutionSchedule::createStageExecution)
+                .collect(toImmutableSet()));
+    }
+
+    @VisibleForTesting
+    SequentialExecutionSchedule(Collection<StageExecution> stages)
+    {
+        fragmentDependency = new DefaultDirectedGraph<>(new FragmentsEdgeFactory());
+        fragmentTopology = new DefaultDirectedGraph<>(new FragmentsEdgeFactory());
+        stagesByFragmentId = stages.stream()
+                .collect(toImmutableMap(stage -> stage.getFragment().getId(), identity()));
+        extractDependencies(
+                stages.stream()
+                        .map(StageExecution::getFragment)
+                        .collect(toImmutableList())).stream()
+                // start non-lazy stages
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+        // start stages without any dependencies
+        fragmentDependency.vertexSet().stream()
+                .filter(fragmentId -> fragmentDependency.inDegreeOf(fragmentId) == 0)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+    }
+
+    @Override
+    public Set<PipelinedStageExecution> getStagesToSchedule()
+    {
+        schedule();
+        return activeSources.stream()
+                .map(StageExecution::getPipelinedStageExecution)
+                .collect(toImmutableSet());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        // dependency graph contains both running and not started fragments
+        return fragmentDependency.vertexSet().isEmpty();
+    }
+
+    @Override
+    public synchronized ListenableFuture<Void> getRescheduleFuture()
+    {
+        return rescheduleFuture;
+    }
+
+    @VisibleForTesting
+    void schedule()
+    {
+        removeCompletedStages();
+        unblockStages();
+    }
+
+    @VisibleForTesting
+    DirectedGraph<PlanFragmentId, FragmentsEdge> getFragmentDependency()
+    {
+        return fragmentDependency;
+    }
+
+    @VisibleForTesting
+    Set<StageExecution> getActiveSources()
+    {
+        return activeSources;
+    }
+
+    private void removeCompletedStages()
+    {
+        Set<StageExecution> completedStages = activeSources.stream()
+                .filter(this::isStageCompleted)
+                .collect(toImmutableSet());
+        // remove completed stages outside of Java stream to prevent concurrent modification
+        completedStages.forEach(this::removeCompletedStage);
+    }
+
+    private void removeCompletedStage(StageExecution stage)
```

Comment:
ditto

Reply:
Idk. I thin the current name is more explicit. Stage is actually removed here
```
        fragmentDependency.removeVertex(fragmentId);
        fragmentTopology.removeVertex(fragmentId);
        activeSources.remove(stage);
```